### ISSUE TITLE
<ChartCanvas /> won't re-render when initialDisplay is changed

### ISSUE THREAD

**Issue by gaborluk** [rrag/react-stockcharts#44]
I'm trying to get a `<ChartCanvas />` re-render by changing its `initialDisplay` and `interval` props, however, the component doesn't react to these changes.

For comparison, if I change the `width` property, that does trigger a re-render every time.

Is this a bug, or intended behavior? If the latter, what is the preferred way of rendering a different fixed time interval for display on user interaction?

**Comment by rrag** [comment]
`initialDisplay` is a hack to show a fixed count of periods while rendering initially. So that is intended behavior. I named the property `initialDisplay` to avoid that confusion. If you want to show a different time range after initial render. I suggest you use the 

`this.refs.chartCanvas.setViewRange` where `chartCanvas` is

```jsx
<ChartCanvas ref="chartCanvas" ...
```

This is what I have used in the [demo](http://rrag.github.io/stockcharts-demo/) when you click on the shortcuts for 6m, 1y, all etc

About `interval` it is a bug, and I have not really tested that portion at all. Even in the demo, I did not use it. Will have to revisit that after intra day scale is implemented, so it can support both end of day and intra day.

**Comment by gaborluk** [comment]
Thank you for the lightning-fast reply. The behavior on the demo is exactly what I was looking for, I will implement your `ref` solution.

**Comment by gaborluk** [comment]
@rrag, I searched your [github.io repo](https://github.com/rrag/rrag.github.io) for the use of `setViewRange`, but can't find an instance of it anywhere. Is the use of this method documented anywhere? If so, could you please point me to it?

**Comment by Pinxie** [comment]
Hello,

You have an example here : http://plnkr.co/edit/gist:11c28c8e6612055ba8a7?p=preview (handleBrush function)

It's the plunker for http://rrag.github.io/react-stockcharts/documentation.html#/brush

An "handleBrush" is done via drawing a rectangle with 2 left clicks inside the chart.

**Comment by gaborluk** [comment]
Hello @Pinxie, thank you, but unfortunately this doesn't show how the `left` and `right` coordinates should be obtained, or how the puzzle pieces fit together, as in this case things are being handled by the plugin internally, and in my case I need to manually control the component from the outside.

**Comment by rrag** [comment]
Thanks @Pinxie 

I don't have my computer with the source of the demo right now. This might help.
```js
this.refs.chartCanvas.getDataInfo()
```

that returns an object like [this](https://github.com/rrag/react-stockcharts/blob/master/src/lib/EventHandler.js#L387-L390)

use the `viewData` which is the filtered array which you are viewing now and `fullData`
Using both of these. you can come up with a `left` and `right`. and pass that to the `setViewRange` method

**Comment by gaborluk** [comment]
Thank you, @rrag, I have implemented it using `getDataInfo()`, and now it more or less works. One issue still persists, though, and I believe that's bound to stay with the plugin's current design – but maybe I'm just using it wrong.

Since `idx` values don't necessarily correspond 1:1 with days, one needs to do a conversion between `date` and `idx` values in order to work with days.

```
const fullRangeInDays = (fullData.end.date - fullData.start.date) / (24 * 60 * 60 * 1000);
const fullRangeInIndices = fullData.end.idx - fullData.start.idx;
const dayInIdx = fullRangeInIndices / fullRangeInDays;

const desiredRange = chartRanges[chartRangeIndex].days * dayInIdx;
```

However, now the same conversion ought to be done for the initial value passed into `initialDisplay`, otherwise, should the component render more than once after page load for any reason, the second time it will grab the range values set by `initialDisplay` sans-conversion, and apply the range setting logic specified in a `componentDidUpdate` using basically wrong coordinates.

Unfortunately one can't apply the same conversion logic to the value passed into `initialDisplay`, since at that point `refs` are still empty, and they shouldn't be accessed from `render` anyway.

Your demo takes an approach with the range setter where the fixed point of the range is always `viewData.end.idx`, and with that approach, you don't run into this issue, however, I went for a center-fixed zooming, and with that, this is a problem. Basically, the only difference between the two approaches is that:
- with the `viewData.end.idx` being the fixed point, the range is basically fixed to the **same coordinate** both at initial render time _and_ with any subsequent render, while:
- with the center coordinate being the fixed point, the range is first fixed to `viewData.end.idx` on initial render, but is being adjusted to the view's center point with every subsequent render, and since the conversion doesn't happen in the first case but does happen in every subsequent case, a subsequent render will move the view range away from the latest data points (near `viewData.end.idx`), which is bad user experience.

Here's my code to make you better understand the issue:
```
  componentDidUpdate () {
    this.applyDesiredViewRange();
  }
```
```
  applyDesiredViewRange () {
    if (this.refs.chartCanvas) {
      const {
        chartRangeIndex
      } = this.props;

      const {
        fullData, // data from all time
        viewData  // data within view
      } = this.refs.chartCanvas.getDataInfo();

      /**
       * We need to do a conversion here between proper Date values (.date) and
       * arbitrary index values (.idx) contained in the data info objects to
       * calculate the desired start and end coordinates for our view range.
       */
      const fullRangeInDays = (fullData.end.date - fullData.start.date) / (24 * 60 * 60 * 1000);
      const fullRangeInIndices = fullData.end.idx - fullData.start.idx;
      const dayInIdx = fullRangeInIndices / fullRangeInDays;

      const desiredRange = chartRanges[chartRangeIndex].days * dayInIdx;

      if (desiredRange === Infinity) {
        this.refs.chartCanvas.setViewRange(fullData.start.idx, fullData.end.idx);
      } else {
        const desiredMidIndex = viewData.start.idx + (viewData.end.idx - viewData.start.idx) / 2;
        const desiredStartIndex = desiredMidIndex - desiredRange / 2;
        const desiredEndIndex = desiredMidIndex + desiredRange / 2;

        // Note: 10 indices is the smallest React Stockcharts can handle
        this.refs.chartCanvas.setViewRange(Math.min(desiredEndIndex - 10, desiredStartIndex), desiredEndIndex);
      }
    }
  }
```
```
  render () {
    ...
    return (
      <ChartCanvas ref='chartCanvas' initialDisplay={chartRanges[chartRangeIndex].days} ... />
    ...
```
(Note how `chartRanges[chartRangeIndex].days` is not converted to `idx` because it can't be where it is.)
<hr>
This is currently not an issue for me anymore, as it is acceptable on this particular project to display the full range initially, and with that, the problem naturally doesn't manifest itself. However, should I want to initiate the chart with, say, a 3-month view range, the issue would immediately be relevant. Unless I'm misusing something, this should be a good reason to reconsider the design a bit – maybe a declarative solution would be possible where one simply passes props to the component the React way, and it updates itself with the proper view range. In any case, thank you for your time and kind help.

**Comment by rrag** [comment]
Excellent explanation of the problem.

I would like to discuss with you to see if there are better ways to expose the interfaces to make the api intuitive.

RE: fixed point being middle approach

I had a similar problem with the zoom which I struggled with for a few days. Zoom out/in is centered on where the mouse pointer is. While the zoom out/in does not have to jump to a fixed duration, may be there are some similarities.

> However, should I want to initiate the chart with, say, a 3-month view range, the issue would immediately be relevant

When I started this, I went with such an approach, where the start and end dates would be sent as  input props. Once the chart is rendered with that initial data, say the user does a zoom/pan, those props no longer represent the true state of the chart. Now when someone resizes the window, and passes a new width/height it was resetting to the initial view port.

To over come this, I had logic to check if only the width/height changed and not the data or the start/end date. I found this kind of logic in the `componentWillReceiveProps` hard to understand and decided against it. There is still this kind of logic there, for some cases.

This is the reason I decided to use `refs.chartCanvas.pushData / alterData` approach to push/edit points. As against the react way of updating the data via props, because I did not know if it was entirely new data, or just data with a few more points added/modified. I could not reliably find if the chart reset the view port.

I need to add support for intra day, and it is probably time for another refactor.
I don't quite like the way the different intervals are handled, but they were the best I could come up with at that time. I am close to releasing 0.4 with a whole bunch of new indicators, once that is done. I will look at this again. As you can imagine I am biased, but open to ideas and suggestions. Appreciate help in making this better.

**Comment by gaborluk** [comment]
I will address a couple of your points where I had a thought that just might change the game, but I'm painfully aware that I'm painfully unaware of the hundreds of problems you've already gone through that are very specific to this app, so it's difficult to be truly helpful here.

> When I started this, I went with such an approach, where the start and end dates would be sent as input props. Once the chart is rendered with that initial data, say the user does a zoom/pan, those props no longer represent the true state of the chart.

You'll probably have a good reason why this wouldn't work, but I can't help thinking that this should just be a regular _do interaction -> update state -> pass down props_ story in a perfect world. These days I use React with Redux, and in this particular case, my thinking would be: on zoom/pan, dispatch an `action` with the intention of modifying the current state – a `reducer` handles that action, and updates the state. Once the state has been changed, it's the classic tale of passing down all the props and letting React do its job. The start and end props shouldn't need to be static at all, they should represent the current state at all times. They would sit in the `store`, surviving other, irrelevant interactions like a window resize.

> I did not know if it was entirely new data, or just data with a few more points added/modified.

I'm not entirely sure about this one, but [this suggests to me](http://facebook.github.io/react/docs/tags-and-attributes.html#svg-elements) that it shouldn't matter – React should intelligently diff SVG, so passing in all the data with just some modifications should result in expected and performant behavior. By the sound of it, you're essentially diffing the data manually in a way.

In any case, I think that if you make an exception to the React way of doing things once, then it's like domino: you'll need to keep making exceptions, and eventually you're back to manual state handling. Also, when it comes to providing an intuitive API, deviating from the expected is bound to make heads hurt. For what it's worth, my preferred way of handling my particular problem from above would look something like this:
```javascript
<ChartCanvas range={chartRanges[chartRangeIndex].days} rangeUnit='days' rangeMode='middle'  ... />
```
The component should know its state, including the current coordinates, internally. With no existing state (initial render), it would snap the view to the latest data point, otherwise it would consider `rangeMode` and render a view range for me based on the current `viewData` in state, and the dynamic value I pass down in `range`. Ideally, I should never need to manually tweak view coordinates directly.

If I'm really off the mark with something, or everything, don't hesitate to point it out.

**Comment by rrag** [comment]
I am afraid it is not that simple. I think I did not explain well before.

say on initial render you have this

```jsx
<ChartCanvas range={chartRanges[chartRangeIndex].days} rangeUnit='days' rangeMode='middle'
data={...} width={...} height={...}
...>
    <Chart id={0} ...>
        ...
    </Chart>
    <Chart id={1} ...>
        ...
    </Chart>
</ChartCanvas>
```

Since it is initial render there is no state of the component, and would copy some of the props `data`, `range` and some others, but not `width` since user interaction will not change it. In addition to these there will be derived fields, like `xScale` `yScale` for each `Chart`

The reason for copying some of these to state is that, they will change when the user interacts, pan/zoom and now the range will be different. start & end domain values will be different than initial render. Also the overlays (like EMA, MACD etc) have to be calculated and more attributes added to the data.

So after initial render, the chart has some state. and with user interaction, the `range` which is passed in the prop is not the range you see on the chart.

Now say you add `<Chart id={2} ...>`. and change nothing else. You would expect to see the additional chart for the same range since the end of last interaction. In this case I have to ignore the incoming props, and not recalculate the state. To do that 
``js
componentWillReceiveProps(nextProps) {
  // compare this.props.range and nextProps.range to see if they are same
}
``

Now say you change the `width` and `height` there is no need to compare the props and prevProps, since `width` is not stored in state, the incoming `width/height` can be used all the time.

Look at the [updating data](http://rrag.github.io/react-stockcharts/documentation.html#/updating_data) example, if the last tick is visible, and a new data point is added. The user would expect to see that data point. but if they pan the chart to the left where the last tick is not visible, and now if a new data point is added. they should not be taken back to the end to see the newly added last tick. This is the reason why in the updating data example the `alterData` `pushData` methods are used.

Holding state in a single place solves many problems. Unfortunately for someone who uses this library. the state is stored in 2 places, one in their application, and a state of the chart inside the chart itself. This makes sense since the user interaction and other derived attributes are in the scope of this library itself.

Since this conversation began, I am thinking if there is a way to send the state of the chart back to the user every time after any interaction, and not hold any state at all. More I think about it, I feel that will introduce complexity.

**Comment by gaborluk** [comment]
That was a very clear explanation, thank you, @rrag. I can see the problem now, and it's really not on the trivial side of the scale. I can't promise I'll have a lot of time to think about this, but I might do, and get back to you in case I come up with something worthy of mention.

**Comment by rrag** [comment]
This discussion was very useful, led me to rethink the api. The changes are made in `next` branch to be released as `0.4`.

`alterData(...)`,  `pushData(...)`, `setViewRange(...)` are no longer part of the api, just use the react way of updating the props.

Early version of the code examples in 
http://rrag.github.io/react-stockcharts-next/documentation.html#/updating_data
http://rrag.github.io/react-stockcharts-next/documentation.html#/brush

plunker/gist/block not updated yet
All the documentation is not updated. Still working on them

Thanks for your input @gaborluk appreciate it.

**Comment by gaborluk** [comment]
This is good news, many thanks for your work, @rrag.
